use stream id instead of stream name for Filter (and some other related places)

evykassirer opened the pull request:
This PR should have no user-facing changes except for being able to enter`stream: $stream_id` in search input (you would enter `stream: $stream_name` before and also after this change).

This is a very challenging refactor to find bugs for. I found several bugs through unit tests and futzing around in the dev environment, but I don't feel like there's any particularly good way to guarantee that all possible things that could break have been tested since the filter code is used for so many things. Seeking advice on what to do about that.


---

Two things I'm considering:

* ~~Looking for help with what we do in `slug_to_id_string` when neither id nor name is present. Maybe what's already written there is fine? It similar to what we did before~~
* Maybe make a prep commit removing the use of `_sub`?

timabbott commented on web/src/compose_reply.ts:
```diff
@@ -86,7 +86,10 @@ export function respond_to_message(opts: {
+                first_operator === "stream" &&
+                !stream_data.is_subscribed(Number.parseInt(first_operand, 10))
+            ) {
                 compose_actions.start({
```
Should this logic just be using `narrow_state.stream_id`? That might be a prep refactor to reduce the madness here.

(I have a vague recollection of us discussing this weird block of code previously).

evykassirer replied:
Ah yeah that would work very well, thanks!

timabbott commented on web/src/filter.ts:
```diff
@@ -291,7 +290,6 @@ const USER_OPERATORS = new Set([
 
 export class Filter {
     _terms: NarrowTerm[];
-    _sub?: StreamSubscription | undefined;
```
What's the thinking around removing this?

Maybe it'll become clear after further reading, but a reason this might be useful is that its presence would confirm that the stream ID is valid and maps to a valid stream.

evykassirer replied:
It seemed like calling `stream_id()` did what was needed. I guess it doesn't confirm the stream id is valid, but it feels simpler to just work with the stream id than to have an extra variable storing the sub. I'm open to keeping it (putting it back) though!

timabbott replied:
The extra lookups might be important in the context of performance optimization of very hot code paths, like `message_matches_search_term`, but we have an issue for optimizing that, so maybe we can deal with that later.

evykassirer replied:
Okay, I'll leave it as is for now then

timabbott commented on web/src/filter.ts:
```diff
@@ -1502,8 +1526,8 @@ export class Filter {
-        return this.has_operand("channel", stream_name) && this.has_operand("topic", topic);
+    has_topic(stream_id: string, topic: string): boolean {
+        return this.has_operand("channel", stream_id) && this.has_operand("topic", topic);
     }
```
Given that `filter.has_topic` only has one caller outside tests, I think it might be more reasonable to have this accept a `stream_id`  integer, and do the translation to a `string_id_str` internally.

This change would be good for consistency; it's nice if every `stream_id` variable is an integer.

evykassirer replied:
Yeah that sounds good, thanks!

timabbott commented on web/src/stream_data.ts:
```diff
@@ -267,20 +278,33 @@ export function slug_to_name(slug: string): string {
+            return stream.stream_id.toString();
+        }
+        // TODO(evy): what should we do here?
+        return slug;
```
I don't think you need any new code here. The case discussed in this comment is that where the stream name is "4-horsemen", which might happen to match the pattern for modern URLs. The intended behavior is to fall through to the logic below for handling the pre-2018 URLs, which do not start with the stream ID.

(This fallthrough was about pre-2018 URLs that appear to be a modern URL because the stream name starts with the pattern `4-`)

evykassirer replied:
So you're saying remove this whole block?

timabbott commented on web/src/stream_data.ts:
```diff
@@ -267,20 +278,33 @@ export function slug_to_name(slug: string): string {
+    }
+    // TODO(evy): what should we do here?
     return slug;
 }
```
I think this should probably return `undefined`, and callers should treat this as an invalid stream reference, the same as it would if the integer ID returned after this parsing didn't exist.

gnprice replied:
For this function I recommend taking a look at the mobile code. We converted this area (and basically the rest of the app's internal data structures) to numeric stream IDs a couple of years ago, and the conversion corresponding to this function took some thought but I'm pretty happy with the result.

Here's the version in zulip-mobile:
https://github.com/zulip/zulip-mobile/blob/2217c858e207f9f092651dd853051843c3f04422/src/utils/internalLinks.js#L66-L108

Or for completeness here's the latest version, in zulip-flutter — but it's basically the exact same code translated to Dart:
https://github.com/zulip/zulip-flutter/blob/30bdcc0fd5b949270fc76ddf173ccb4952e287a2/lib/model/internal_link.dart#L262-L296

Two key points are: it can return a nice numeric ID instead of a string; and the pre-2018 case needs a bit of new logic at the end of the function.

gnprice replied:
Oh and here's test cases written for exercising the corner cases of this interaction between the pre-2018 form and the modern form:
https://github.com/zulip/zulip-mobile/blob/2217c858e207f9f092651dd853051843c3f04422/src/utils/__tests__/internalLinks-test.js#L333-L354

evykassirer replied:
Thanks, this was helpful! I made a few changes to match what mobile is doing.

timabbott commented on web/src/hash_util.ts:
```diff
@@ -288,7 +288,7 @@ export function validate_group_settings_hash(hash: string): string {
 export function decode_stream_topic_from_url(
     url_str: string,
-): {stream_name: string; topic_name?: string} | null {
+): {stream_id: string; topic_name?: string} | null {
```
Call this `stream_id_string` or return the integer, whichever seems cleaner for how this is used.

evykassirer replied:
Returning a number seems good

timabbott commented on web/src/stream_settings_ui.js:
```diff
@@ -888,7 +888,7 @@ export function view_stream() {
     if (row_data) {
         const stream_narrow_hash =
-            "#narrow/stream/" + hash_util.encode_stream_name(row_data.object.name);
+            "#narrow/stream/" + hash_util.encode_stream_id(row_data.object.stream_id);
```
I'm pretty sure that `stream_id` value is an `integer`; so `hash_util.encode_stream_id` should expect an integer (it only has two callers, and generally using integer IDs seems like this simpler/more natural API if we're tiebreaking).

You can test this code path by clicking the "preview channel" button in stream settings, I think.

evykassirer replied:
Ah yes, thanks! Changed `encode_stream_id` to take a number

timabbott commented on web/src/message_view.js:
```diff
@@ -1065,10 +1066,9 @@ export function render_message_list_with_selected_message(opts) {
+function activate_stream_for_cycle_hotkey(stream_id) {
     // This is the common code for A/D hotkeys.
-    const filter_expr = [{operator: "channel", operand: stream_name}];
+    const filter_expr = [{operator: "channel", operand: stream_id.toString()}];
```
I feel like a lot of code paths calling `show` .

Would it make sense to have `operand` being an integer be a valid encoding?

```$ git grep operand | grep toString
web/src/filter.ts:            return message.id.toString() === operand;
web/src/filter.ts:                operand = operand.toString().toLowerCase();
web/src/filter.ts:                operand = operand.toString().toLowerCase();
web/src/filter.ts:                operand = operand.toString().toLowerCase();
web/src/filter.ts:            return sign + operator + ":" + Filter.encodeOperand(term.operand.toString());
```

evykassirer replied:
Did you mean to say "I feel like a lot of code paths call `toString`"?

I meant to talk a bit about this, as I was thinking about it. I agree converting between string and integer in several places is not great.

However, it feels a bit hard to add an integer operand, because `channel:Denmar` is a parseable search term, and we only change the channel operand to a number once the stream name is valid -- so channel can have string and (currently stringified) number operands. I don't really like that this is the case, and maybe there's a nicer way to parse terms that's more elegant, but it feels like supporting integer channel id operands would be a much bigger change.

timabbott replied:
Yeah, that's what I meant.

Note that `filter.parse` has to handle a raw string, but there's nothing requiring the fact that you can type `"channel:Denmark" to say anything about the required format of processed `NarrowTerm` structures that we use; it would be totally reasonable to have parsing turn channel names into either an ID or an error, without that needing to contaminate these internal APIs).

evykassirer replied:
`search_suggestion.ts` uses `NarrowTerm`s that we parsed, and that has to include partially typed out channel names which are part of how we give channel suggestions.

timabbott replied:
If there's only one place that needs a different format, that code path could do whatever logic is required to produce correct output; I don't see why that should have a strong constraint on the format of internal data structures.

evykassirer replied:
This seems worth considering, but as a followup. I think it would involve separating `NarrowTerm` into `SearchTerm` (all string) and `NarrowTerm` (only after validation, where channels always have number operands).

timabbott replied:
Yeah, that seems like a reasonable way to proceed here.

evykassirer replied:
Cool, added this to my list of non-urgent and somewhat-involved followups

timabbott reviewed (commented):
OK did a partial read (didn't really get into the tests), and I think these are the main questions.

timabbott commented on web/tests/narrow_state.test.js:
```diff
@@ -59,16 +59,17 @@ test("stream", () => {
     const public_terms = narrow_state.public_search_terms();
     assert.deepEqual(public_terms, expected_terms);
-    assert.equal(narrow_state.search_string(), "channel:Test topic:Bar yo");
+    assert.equal(narrow_state.search_string(), "channel:15 topic:Bar yo");
```
Is `narrow_state.search_string` a function we should be garbage-collecting at this point? I'm not sure it has any callers.

evykassirer replied:
Great find! Yeah looks like no callers. I'll add a prep commit removing that

timabbott commented on web/tests/narrow_state.test.js:
```diff
@@ -214,7 +216,7 @@ test("set_compose_defaults", () => {
 
     stream_data.add_sub({name: "ROME", stream_id: 99});
-    set_filter([["stream", "rome"]]);
+    set_filter([["stream", "99"]]);
```
I think we want the pattern for these to look more like `rome_stream.stream_id``, possibly with a `toString` depending on how other comments are resolved. Opaque integers like this make the code hard to read.

Looks like a preexisting issue with this test more broadly, though, two lines down I see a hardcoded 99.

evykassirer replied:
Strings for filter terms and numbers for `set_compose_defaults` is the correct typing. But yeah I'm doing a pass right now to not use opaque string integers. (I had a `// TODO(evy): make these more like `channel:${dev_help.stream_id}` below but I wasn't sure how important it was and your comment encouraged me to go ahead and do that)

evykassirer commented on web/src/narrow_banner.ts:
```diff
@@ -184,7 +185,7 @@ export function pick_empty_narrow_banner(): NarrowBannerData {
             page_params.is_spectator &&
             first_operator === "channel" &&
-            !stream_data.is_web_public_by_stream_name(first_operand)
+            !stream_data.is_web_public_by_stream_id(Number.parseInt(first_operand, 10))
```
This can also use `narrow_state.stream_id`

evykassirer commented on web/src/narrow_banner.ts:
```diff
@@ -264,7 +265,7 @@ export function pick_empty_narrow_banner(): NarrowBannerData {
             break;
         case "channel":
-            if (!stream_data.is_subscribed_by_name(first_operand)) {
+            if (!stream_data.is_subscribed(Number.parseInt(first_operand, 10))) {
```
This can also use `narrow_state.stream_id`

evykassirer commented:
@timabbott I believe I've addressed all comments!

gnprice commented on web/tests/stream_data.test.js:
```diff
@@ -166,16 +162,21 @@ test("basics", () => {
 
-    assert.equal(stream_data.slug_to_name("99-whatever"), "99-whatever");
-    assert.equal(stream_data.slug_to_name("99whatever"), "99whatever");
+    // invalid formats
+    assert.equal(stream_data.slug_to_stream_id("25-or-6-to-4"), undefined);
+    assert.equal(stream_data.slug_to_stream_id("2something"), undefined);
+    assert.equal(stream_data.slug_to_stream_id("99-whatever"), undefined);
+    assert.equal(stream_data.slug_to_stream_id("99whatever"), undefined);
```
The first and third of these are validly formatted in the new style: they mean a stream/channel with ID 25 or 99 respectively.

All four are validly formatted in the old style: they mean streams with those names.

evykassirer replied:
Currently `slug_to_stream_id` just gets a valid stream id, and `decode_operand` is doing the rest of the work, including returning the full operand (i.e. the name) if there's no valid stream id found.

`25-or-6-to-4` and `99-whatever` returned `undefined` because `25` and `9` weren't valid ids. I could do a followup test adding streams with those ids and show that it returns the id then?

I'm thinking maybe I should keep `slug_to_name` and these tests cases, and add a new function `slug_to_id` that is used by `slug_to_name`.

gnprice replied:
> returning the full operand (i.e. the name) if there's no valid stream id found.

I see. So I guess then the behavior question comes down to: if the stream operand we're looking at
* doesn't correspond to any known stream using either format,
* but does have the shape consistent with the post-2018 format (i.e. it matches `/^(\d+)(?:-.*)?$/`),
* then should we interpret it according to the post-2018 format or the pre-2018 format?

I think the preferable answer is we should interpret it with the post-2018 format. It seems like it'll be way more likely that an operand `99-whatever` is a post-2018 link to an (unknown) stream with ID 99 than that it's a pre-2018 link to an (unknown) stream with name "99-whatever". Both because 2018 was a long time ago now, and because it'd be kind of a coincidence for the stream name to have the shape `/^(\d+)(?:-.*)?$/` anyway.

The behavior in this version is effectively that we interpret such an operand with the pre-2018 format — that's what it means to take the whole operand as the stream name, instead of taking the leading `(\d+)` group as the stream ID.

(Perhaps relatedly, I'd say 25 and 9 are always "valid stream IDs", regardless of what streams the client may or may not know about, or what streams may or may not exist on the server.)

gnprice replied:
One bonus of that approach is that it means you can have `decode_operand` always return (a stringification of) a numeric stream ID when the operand is `stream`/`channel`. Then everything that consumes such operands can know that's the case, and not have to ever wonder if it's a stream name rather than a stream ID.

evykassirer replied:
> I'm thinking maybe I should keep `slug_to_name` and these tests cases, and add a new function `slug_to_id` that is used by `slug_to_name`.

I tried this and liked it, and will push that shortly.

If we're just showing "unknown channel" without more copy, then I don't think it matters what we return when we can't identify a stream id. Does that sound right?

gnprice replied:
> If we're just showing "unknown channel" without more copy, then I don't think it matters what we return when we can't identify a stream id. Does that sound right?

Yeah, it might not matter in that case, depending on what else the code that consumes these results is doing with the information. (I haven't looked.)

gnprice replied:
If it truly doesn't matter, though, it'd be best to not propagate all the mess of the full slug — better to either return the presumed stream ID, or some fixed value like "" or null or undefined.

Passing around the full slug when it doesn't mean anything will be an invitation for bugs, as future people working on other code assume it's supposed to potentially mean something and propagate new complicated logic trying to interpret it.

evykassirer replied:
>  depending on what else the code that consumes these results is doing with the information. (I haven't looked.)

Looking -- `slug_to_name` is only called by `decode_operand` which is only called by `parse_narrow`, which is called in two places:

*`hashchange.do_hashchange_normal`, and the parsed terms are sent to `message_view.show`
  * I haven't looked at this in depth, but it seems to be the stuff we've been focusing on so far, which is the navbar header and the message feed contents
* `hash_util.decode_stream_topic_from_url`, only used in `copy_and_paste.try_stream_topic_syntax_text`
  * Here it makes sense to return `null` from `decode_stream_topic_from_url` if we can't find a valid stream id, so I'll change that but it doesn't make a difference for this conversation.

I double checked that what we return doesn't seem to affect normal channel narrowing or search bar or reply behavior or anything, so I think we can return whatever we want here.

So here's what we could do:

* Return a valid stream id, if we can find one from a prepended number on the slug (new format) or from the full slug as a valid stream name (old format)
* return an empty string (or undefined?) from `slug_to_name` (and therefore `decode_operand`) if we can't find a valid stream id

If we did this, it would be pretty much what I had before, so I'd go back to not having `slug_to_name` and only `slug_to_stream_id` which can return `undefined`, and have `decode_operand` call that directly. Probably the main difference would be defaulting to an empty string here, instead of falling back to returning the `operand`:

```js
    if (util.canonicalize_stream_synonyms(operator) === "stream") {
        return stream_data.slug_to_id(operand) ?? "";
    }
```

evykassirer replied:
I'm going to do this

gnprice commented on web/tests/stream_data.test.js:
```diff
@@ -166,16 +162,21 @@ test("basics", () => {
+    assert.equal(stream_data.slug_to_stream_id("2"), 2);
+    // we still get 2 because it's a valid stream id
+    assert.equal(stream_data.slug_to_stream_id("2-whatever"), 2);
+    // invalid stream id
+    assert.equal(stream_data.slug_to_stream_id("999-social"), undefined);
     // legacy
-    assert.equal(stream_data.slug_to_name("25-or-6-to-4"), "25-or-6-to-4");
```
This is a valid thing that can appear in a legacy link, so it should continue to work. That means it should continue to resolve to a stream with this name — if such a stream exists.

That's the "311-help" example in the mobile tests:
https://github.com/zulip/zulip-mobile/blob/2217c858e207f9f092651dd853051843c3f04422/src/utils/__tests__/internalLinks-test.js#L333-L354

The "311-" and "311" examples are related edge cases that would also be good to test.

evykassirer replied:
Should `25-or-6-to-4` return `"25"` as a stream id? Or `"25-or-6-to-4"` as a full stream name because 25 isn't a valid stream id? I just switched to doing the former, but that will change this test case from how it was before. This is for an operand for the stream operator, and I think actually keeping it as a full string for cases when we don't recognize the id makes more sense?

evykassirer replied:
I think the "311" test case is `assert.equal(stream_data.slug_to_name("2"), "2");`

I can add a `assert.equal(stream_data.slug_to_name("2-"), "2");`

(these are both for valid stream ids)

evykassirer replied:
Yeah I agree! I was working on changing it something like that, I'll just push it and add some test cases to show what's going on.

gnprice replied:
Followup thread below: https://github.com/zulip/zulip/pull/31299#discussion_r1765739302

evykassirer commented on web/src/stream_data.ts:
```diff
@@ -259,29 +276,29 @@ export function slug_to_name(slug: string): string {
+    // "New" (2018) format: ${stream_id}-${stream_name} .
+    const match = /^(\d+)(-.*)?$/.exec(slug);
+    const newFormatStreamId = match ? Number.parseInt(match[1]!, 10) : null;
+    if (newFormatStreamId && stream_info.get(newFormatStreamId)) {
```
Ah yes thanks!

evykassirer commented on web/src/stream_data.ts:
```diff
@@ -259,29 +276,29 @@ export function slug_to_name(slug: string): string {
+        return stream.stream_id;
+    }
+
+    // Unexpected shape, or the old shape and we don't know of a stream with
+    // the given name.
+    return undefined;
```
Ah okay, so we do want to return ids even if they might be invalid. Okay I think that's pretty easy to do here!

evykassirer replied:
I'm actually unsure if we want do this -- see my comment below in the test file.

evykassirer replied:
Navigating to `http://localhost:9991/#narrow/stream/999` either way throws an error, so I just need to separately handle this case. But I think it would be nicer to show the full slug when saying we don't know what the channel is.

gnprice replied:
If the link is in the post-2018 form (which it is by far most likely to be these days), then the slug as a whole is a bit of a computery made-up format, not a user-chosen name — like "2-general" rather than "general". So I'm not sure it's helpful to foreground the slug string anyway.

In particular if we mention the slug as if it's the channel name, that's probably wrong (because the link was probably generated since 2018, in the post-2018 form). If the user followed a link with that operand "2-general" and they lack access, it's probably that there's a channel with ID 2, not a channel with name "2-general". So saying something like "you don't have access to channel '2-general'" is going to be misleading. The best option is probably like "that channel doesn't exist or you lack access", without attempting to say the channel's name.

evykassirer replied:
Okay cool, so I'll stick to just saying plainly "Unknown channel" then, since that's simple and less likely to be misleading or incorrect.

timabbott commented:
This might need a rebase following #30509. I'm also not sure if the above conversation about handling corner cases properly is complete.

evykassirer commented:
Rebased and made an update about that corner case and added some tests to show what's going on there.

I think avoiding `toString()` and `parseInt` would take enough refactoring that it shouldn't be part of this PR, but I can do it as a followup.

evykassirer commented:
@timabbott CI is passing, ready for your review!

timabbott commented on web/src/stream_data.ts:
```diff
@@ -271,9 +282,9 @@ export function slug_to_stream_id(slug: string): number | undefined {
+    const match = /^(\d+)(?:-.*)?$/.exec(slug);
     const newFormatStreamId = match ? Number.parseInt(match[1]!, 10) : null;
-    if (newFormatStreamId && stream_info.get(newFormatStreamId)) {
+    if (newFormatStreamId !== null && stream_info.get(newFormatStreamId)) {
```
Is it intentional that this reworks the new `slug_to_stream_id` function? Might be helpful to have that in its own commit.

evykassirer replied:
Yeah this came from previous review comments:
* https://github.com/zulip/zulip/pull/31299#discussion_r1722270902
* https://github.com/zulip/zulip/pull/31299#discussion_r1722276031

Separating it into its own commit sounds good.

timabbott commented on web/src/stream_data.ts:
```diff
@@ -177,6 +179,20 @@ export function get_sub(stream_name: string): StreamSubscription | undefined {
+    const stream_id = Number.parseInt(stream_id_string, 10);
+    const stream = stream_info.get(stream_id);
+    return stream;
+}
```
These should be in the other order: Usually best to call a function you've already defined. 

Also, I think the naming scheme of having the one that can fail with `assert_valid` in its name, rather than `possible` in the name of the one that can return undefined, would be a lot more consistent with other function names.

evykassirer replied:
What about `get_valid_sub_by_id_string`? It's a bit shorter and more readable in my opinion.

timabbott replied:
I don't like it as much (it doesn't speak as clearly that the caller is claiming the ID string is valid, and thus we expect to be guaranteed to get a sub back; it sounds like we're asking for a "valid sub" instead) but we can merge this and then bikeshed in #frontend.

evykassirer replied:
https://chat.zulip.org/#narrow/stream/6-frontend/topic/naming.20functions.20that.20don't.20return.20undefined/near/1944622

evykassirer commented:
@timabbott addressed your two comments

timabbott commented:
This feels like a mergeable checkpoint, merged, thanks @evykassirer! There's definitely some follow-up work to do on this, and maybe we'll find some exceptions when it goes to chat.zulip.org later today.

gnprice commented on web/tests/stream_data.test.js:
```diff
@@ -168,19 +165,37 @@ test("basics", () => {
+    assert.equal(hash_util.decode_operand("channel", "99-whatever"), "");
 
-    // invalid formats
     assert.equal(stream_data.slug_to_stream_id("25-or-6-to-4"), undefined);
```
I'm not seeing a test that reflects the outcome of this thread above:
https://github.com/zulip/zulip/pull/31299#discussion_r1722289142

This line where the answer is "undefined" comes from the recent 05e58de3f5f507b23d287e678e96879937ad29b7 (#31575), which was split out from this PR. This format is still valid and should still work, when a stream named "25-or-6-to-4" exists (well, is known) and one with ID 25 doesn't.

gnprice replied:
(I think the new implementation does handle that case correctly — there should just be a test showing that.)

evykassirer replied:
There are these two tests

```js
    assert.equal(stream_data.slug_to_stream_id("25-or-6-to-4"), undefined);
    assert.equal(hash_util.decode_operand("channel", "25-or-6-to-4"), "");
```

The part where we `return operand;` because there's no stream id is a function of `hash_util.decode_operand`, not anything in `stream_data`. Did you want a test added for `hash_util` for this? Or something else?

gnprice replied:
Right, those are the test lines I was commenting on :slightly_smiling_face: (or one of them is).

This format is still valid. That means that `slug_to_stream_id` should be able to return an ID for it.

Currently I think we don't have any tests that would fail if this bit of the implementation:
```js
    const newFormatStreamId = match ? Number.parseInt(match[1]!, 10) : null;
    if (newFormatStreamId !== null && stream_info.get(newFormatStreamId)) {
        return newFormatStreamId;
    }
```

were edited to instead look like:
```js
    if (match) {
        // The slug is in the new format.
        const streamId = Number.parseInt(match[1]!, 10);
        if (!stream_info.get(streamId)) {
            // The stream ID isn't for a known stream.
            return null;
        }
        return streamId;
    }
```

Logic like that would look pretty reasonable on the surface, but would break old links where the stream name happens to have the same shape as a new-style slug. So we should have tests that would detect that breakage.

As I said in the linked thread, that's what the "311-help", "311-", and "311" cases in these mobile tests are testing:
https://github.com/zulip/zulip-mobile/blob/2217c858e207f9f092651dd853051843c3f04422/src/utils/__tests__/internalLinks-test.js#L333-L354

evykassirer replied:
I believe that's what these tests are testing (earlier in the file)? Sorry if I'm missing something still.

```js
    assert.equal(stream_data.slug_to_stream_id("2-social"), 2);
    assert.equal(stream_data.slug_to_stream_id("2"), 2);
    assert.equal(stream_data.slug_to_stream_id("2-whatever"), 2);
    assert.equal(stream_data.slug_to_stream_id("2-"), 2);
```

evykassirer replied:
So you're looking for a test that has a stream with a name that starts with a number?

Like maybe after `assert.equal(stream_data.slug_to_stream_id("25-or-6-to-4"), undefined);` adding a stream with the name `"25-or-6-to-4"` and confirming that it identifies that stream. Am I understanding now?

gnprice replied:
Right, exactly.

Then as a supplement to that, it would also be good to have a stream with a numeric name like "25" (or some other number), and check that `slug_to_stream_id` returns the ID of that stream (not undefined, and not the number 25).

And as a further case (or rather pair of cases), check that if the name is like either "25" or "25-or-6-to-4" and there's *both* a stream with ID 25 and a different stream that has the whole slug as its name, the one with the ID wins.

evykassirer replied:
Ok great, that makes sense. Here's a PR with that: #31677

evykassirer commented on web/src/message_view_header.ts:
```diff
@@ -131,11 +131,15 @@ function get_message_view_header_context(filter: Filter | undefined): MessageVie
 export function colorize_message_view_header(): void {
     const filter = narrow_state.filter();
-    if (filter === undefined || !filter._sub) {
+    let current_stream;
```
Made a followup commit for this

evykassirer commented on web/src/message_fetch.ts:
```diff
@@ -340,7 +340,25 @@ export function load_messages(opts: MessageFetchOptions, attempt = 1): void {
+                    // If the sub is undefined, we'll get an error which is handled
+                    // later by showing a "this channel does not exist or is private"
+                    // notice.
+                    operand: sub?.name,
```
This caused a bug (https://chat.zulip.org/#narrow/stream/464-kandra-js-errors/topic/ZodError.3A.20.5B)

evykassirer replied:
#31728
